File: src/quant_system/api/schemas/market_cross_section.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


class MarketCrossSectionMetaResponse(BaseModel):
    provider: Literal["futu"]
    symbol: str
    currency: Literal["USD"]
    timezone: Literal["America/New_York"]
    as_of: str
    adjustment: Literal["qfq"]
    provenance: Literal["futu", "futu_cache"]


class MarketCrossSectionReturnsResponse(BaseModel):
    week_pct: float
    month_pct: float
    ytd_pct: float


class MarketCrossSectionHistoryPointResponse(BaseModel):
    date: str
    close: float
    indexed_return_pct: float


class MarketCrossSectionRowResponse(BaseModel):
    symbol: str
    rank: int
    returns: MarketCrossSectionReturnsResponse
    volatility_pct: float
    max_drawdown_pct: float
    history: list[MarketCrossSectionHistoryPointResponse]
    meta: MarketCrossSectionMetaResponse


class MarketCrossSectionBasketLabelResponse(BaseModel):
    en: str
    zh: str


class MarketCrossSectionResponse(BaseModel):
    """Read-only cross-section over a preset/custom symbol universe.

    Same fail-closed Futu contract as Asia Radar: no sample data, no estimated
    valuation metrics.
    """

    schema_version: Literal["1.0"]
    provider: Literal["futu"]
    as_of: str
    timezone: Literal["America/New_York"]
    fetched_at: str
    provenance: Literal["futu", "futu_cache"]
    basket: str | None
    basket_label: MarketCrossSectionBasketLabelResponse | None
    methodology: dict[str, str]
    rows: list[MarketCrossSectionRowResponse]

File: src/frontend/lib/marketCrossSection.ts
import { apiRequest } from "./apiClient";

export type MarketCrossSectionProvenance = "futu" | "futu_cache";

export type MarketCrossSectionRow = {
  symbol: string;
  rank: number;
  returns: {
    week_pct: number;
    month_pct: number;
    ytd_pct: number;
  };
  volatility_pct: number;
  max_drawdown_pct: number;
  history: Array<{
    date: string;
    close: number;
    indexed_return_pct: number;
  }>;
  meta: {
    provider: "futu";
    symbol: string;
    currency: "USD";
    timezone: "America/New_York";
    as_of: string;
    adjustment: "qfq";
    provenance: MarketCrossSectionProvenance;
  };
};

export type MarketCrossSectionResponse = {
  schema_version: "1.0";
  provider: "futu";
  as_of: string;
  timezone: "America/New_York";
  fetched_at: string;
  provenance: MarketCrossSectionProvenance;
  basket: string | null;
  basket_label: { en: string; zh: string } | null;
  methodology: Record<string, string>;
  rows: MarketCrossSectionRow[];
};

export function getMarketCrossSection(params?: {
  basket?: string;
  symbols?: string[];
}) {
  const search = new URLSearchParams({ provider: "futu" });
  if (params?.basket) {
    search.set("basket", params.basket);
  }
  if (params?.symbols?.length) {
    search.set("symbols", params.symbols.join(","));
  }
  return apiRequest<MarketCrossSectionResponse>(
    `/api/market-cross-section?${search.toString()}`,
    { cache: "no-store" },
  );
}

export type MarketCrossSectionEnvelope = {
  crossSection?: MarketCrossSectionResponse;
  apiError?: string;
};

export async function getMarketCrossSectionSafe(params?: {
  basket?: string;
  symbols?: string[];
}): Promise<MarketCrossSectionEnvelope> {
  try {
    const crossSection = await getMarketCrossSection(params);
    return { crossSection };
  } catch (error) {
    return {
      apiError:
        error instanceof Error
          ? error.message
          : "Market cross-section unavailable",
    };
  }
}
